test/fonts/test_unused_fonts.py

```diff
@@ -0,0 +1,198 @@
+            assert (
+                font_id in used_font_ids
+            ), f"Page {page_num} contains unused font {font_key}"
+
+        if page_num == 1:
+            assert used_font_ids == {2}, "Page 1 should only use F2"
+        elif page_num == 2:
+            assert used_font_ids == {3}, "Page 2 should only use F3"
+        elif page_num == 3:
+            assert used_font_ids == {1, 2, 3}, "Page 3 should use F1, F2, F3"
+
+
+def test_unused_added_font_not_included(tmp_path):
+    pdf = FPDF()
+    pdf.add_font("Roboto", fname=FONT_DIR / "Roboto-Regular.ttf")  # F1
+    pdf.add_font("Roboto-Bold", fname=FONT_DIR / "Roboto-Bold.ttf")  # F2
+
+    pdf.add_page()
+    pdf.set_font("Roboto")
+    pdf.cell(text="Hello")
+
+    output_path = tmp_path / "test.pdf"
+    pdf.output(output_path)
+
+    reader = pypdf.PdfReader(output_path)
+    fonts = reader.pages[0]["/Resources"]["/Font"]
+    assert "F2" not in fonts, "Unused font F2 should not be included"
+
+
+def test_font_set_but_not_used(tmp_path):
+    pdf = FPDF()
+    pdf.add_font("Roboto", fname=FONT_DIR / "Roboto-Regular.ttf")  # F1
+    pdf.add_page()
+    pdf.set_font("Roboto")
+    pdf.add_page()
+    pdf.set_font("Helvetica")
+    pdf.cell(text="Hello")
+
+    output_path = tmp_path / "test.pdf"
+    pdf.output(output_path)
+
+    reader = pypdf.PdfReader(output_path)
+    page = reader.pages[0]
+    # pylint: disable=no-member
+    resources = page.get("/Resources", {})
+    # pylint: enable=no-member
+    page1_fonts = resources.get("/Font", {}) if isinstance(resources, dict) else {}
+    assert not page1_fonts, "Page 1 should have no fonts as none were used"
+
+
+def test_multiple_pages_font_usage(tmp_path):
+    pdf = FPDF()
+    pdf.add_font("Roboto", fname=FONT_DIR / "Roboto-Regular.ttf")  # F1
+    pdf.add_font("Roboto-Bold", fname=FONT_DIR / "Roboto-Bold.ttf")  # F2
+
+    # Page 1: Use F1
+    pdf.add_page()
+    pdf.set_font("Roboto")
+    pdf.cell(text="Page 1")
+
+    # Page 2: Use F2
+    pdf.add_page()
+    pdf.set_font("Roboto-Bold")
+    pdf.cell(text="Page 2")
+
+    output_path = tmp_path / "test.pdf"
+    pdf.output(output_path)
+
+    reader = pypdf.PdfReader(output_path)
+    page1_fonts = reader.pages[0]["/Resources"]["/Font"]
+    page2_fonts = reader.pages[1]["/Resources"]["/Font"]
+
+    # pylint: disable=no-member
+    assert list(page1_fonts.keys()) == ["/F1"], "Page 1 should only have F1"
+    assert list(page2_fonts.keys()) == ["/F2"], "Page 2 should only have F2"
+    # pylint: enable=no-member
+
+
+def test_nested_context_font_usage_after_page_break(tmp_path):
+    pdf = FPDF()
+    pdf.add_page()
+    pdf.add_font("Roboto-Regular", style="", fname=HERE / "Roboto-Regular.ttf")
+    pdf.add_font("Roboto-BoldItalic", style="", fname=HERE / "Roboto-BoldItalic.TTF")
+    pdf.add_font("DejaVuSans", style="", fname=HERE / "DejaVuSans.ttf")
+    pdf.add_font("Garuda", style="", fname=HERE / "Garuda.ttf")
```

Comment:
```suggestion
    pdf.add_font(fname=HERE / "Roboto-Regular.ttf")
    pdf.add_font(fname=HERE / "Roboto-BoldItalic.TTF", style="BI")
    pdf.add_font(fname=HERE / "DejaVuSans.ttf")
    pdf.add_font(fname=HERE / "Garuda.ttf")
```
Mostly stylistic suggestions (not specifying `style` if not needed, and the font name will be deducted from `fname` if not provided), but I think that `style="BI"` was also missing for `Roboto-BoldItalic`.

Reply:
Handled in [`1b48d3e` (#1387)](https://github.com/py-pdf/fpdf2/pull/1387/commits/1b48d3e4a2594b96aa6339daef0a6b364a5cbdfe)